### DISCUSSION TITLE
Allow Macroables to override defined methods

### DISCUSSION THREAD

**Discussion by joelmellon** [laravel/framework#42915] [Ideas] [answered]
I'd like to be able to override defined methods on Macroables.

An example would be overriding the `Str::start()` method to accept an array of values that can "start" a string, and a default if the string doesn't already start with one of the possibilities in the array. eg. I could do:

```php
Str::start('+200', ['+', '-']); // outputs '+200'
Str::start('-200', ['+', '-']); // outputs '-200'
Str::start('200', ['+', '-'], '+'); // outputs '+200' (using the value of the third argument as the "default")
```

Currently, if I declare a macro like so:
```php
Str::macro('start', function($string, $possibilities, $default = null){
    return 'my custom +/- functionality';
});
```
The macro closure doesn't get executed when I call `Str::start('whatever');`. This is because the `Macroable::__call()` and `Macroable::__callStatic()` magic methods aren't called when accessible methods already exist with that name. So we don't get a chance to check for macros before out-of-the-box methods are executed. Rephrased: since it's an actual method, calling `Str::start()` doesn't call `Macroable::__call()` or `Macroable::__callStatic()`.

One option would be to declare all `Str` methods as macros out-of-the-box, but that would involve rewrites for all classes that use the Macroable trait.

I'm happy to make a PR but I guess I need a little inspiration. Any ideas other ways to achieve this functionality without resorting to a rewrite of all public methods of any Macroable classes?

**Comment by henzeb** [comment]
The Facades is also using __callStatic. So a check could be done there, or a proxy method could be added. Instead of` $instance->$method(...)` you could see if the instance is using `Macroable` and if so execute `$instance->__macroCall($method, ...)` which does the actual logic to determine which to execute. 

The problem may lay in the expectations. if I use a function called lowercase to do lowercase, I do not want it to transform it from pascalcase to snakecase. Your example could be a valid usecase but you cannot disallow others without doing extensive inspections, which may or may not impact overall performance.

basically, I like, but I also think it's a bad idea. Macroable was created to add functionality to classes from third parties, not to modify it's behavior.

**Comment by joelmellon** [comment]
There isn't a facade for `Str`. Instead there’s a top-level class alias for `Str` which is what developers are using in their code when they write `\Str::whatever()` or if they `use Str;`.

> Macroable was created to add functionality to classes from third parties, not to modify it's behavior.

That's a pretty strange assumption there. If it wasn't created to "modify it's behavior" then why isn't `Illuminate\Support\Str` declared as `final`? Suggesting that Laravel shouldn't be extended is ridiculous. It's a concept that goes [way back](https://laravel.com/docs/4.2/extending) in its history. It is in fact a _**core**_ concept that its authors have gone to great lengths to make possible in a vast variety of ways with various mechanisms including: IoC container binding, aliases, facades, providers, interfaces which are in place to help with extension compatibility, factories/managers, events, traits, macros of course, form requests, boot and register methods on various classes...most object-oriented design patterns employed by Laravel facilitate "modification" not just "adding functionality".

**Reply by henzeb** [reply]
Well, considering the fact that they didn't build it to modify existing class behavior might ring a bell. It is also technically not possible. the magic functions only get called when there is no actual method. It may also break the open close principle if you could do that.

The only solution to create macros that override existing functionality is using a front, like Facade. For anything that is using a Facade, Laravel could support it, but I don't see the point, as you can configure the facade root and extend the class the regular OOP way already.

just face it. Macroable is not the way to go if you want to extend existing methods.

**Reply by joelmellon** [reply]
You have no idea what you're talking about. Go away.

**Reply by henzeb** [reply]
I'm sorry if my opinion hurt your feelings.

**Reply by joelmellon** [reply]
What a loser. Troll harder.

**Reply by therecluse26** [reply]
Lol so salty

**Reply by Sophist-UK** [reply]
@joelmellon Your comments above are against the Code of Conduct for this repo which state:
* Participants will be tolerant of opposing views.
* Participants must ensure that their language and actions are free of personal attacks and disparaging personal remarks.
* When interpreting the words and actions of others, participants should always assume good intentions.
* Behavior that can be reasonably considered harassment will not be tolerated.


If you don't like what other people are saying or don't like people that disagree with you, please keep your feelings to yourself.

**Comment by joelmellon** [comment, marked as answer]
My current solution works quite well, although it doesn't use macros. It's a more traditional object oriented approach to extension.

I modified the `Str` alias in `config/app.php` to point to my own `Str` class that extends the Illuminate `Str` class. This is the best way I currently know how to "override" `Str` methods. This covers most usage, but not if developers `use Illuminate\Support\Str;` directly. From working on various projects, it’s quite common for developers to do that, so if I want to override a method or two I need to go through the whole project and replace all occurrences of `use Illuminate\Support\Str;`. It also doesn’t cover using the `Str` class alias before Laravel boots, eg. in a config file, so developers need to make sure to use the extended the `Str` class that I created. The `Str` alias isn't registered before the app is booted anyway, so the same limitation is true for macros as well.

Maybe this solution is as good or better than using macros to override methods (if that were possible). It would still be nice to be able to override macros; it would be more plug-and-play.

**Reply by henzeb** [reply]
Well if you want me to 'troll' harder, fine. The things you worry about are not worth any changes in `Macroable` as you can solve them already in other, more suitable ways. What is the point in having a second backdoor if the first is working perfectly?

1. When developers want to use `start` the way you proposed, they will realize at to points that using `use Illuminate\Support\Str` does not work: A, their IDE shows it's not possible, B, their tests will fail. They will automatically realize that they are using the wrong Str class.

2. Instruct your developers to always use the Str alias if you really want them to use your implementation. You might even want to create a rule in a tool like phpcodesniffer or phpmd to enforce that.

3. You cannot change the behavior of a method as such that point 1 would not apply. That would be a violation of the open closed principle. 

4. the proposed changes to `start` might be worthy to implement in laravel itself.

5. If you do not agree with me, please react with arguments instead of calling me a loser...

**Reply by joelmellon** [reply]
Enjoy masturbating to Github threads.

**Reply by Sophist-UK** [reply]
@joelmellon Your comments above are against the Code of Conduct for this repo which state:
* Participants will be tolerant of opposing views.
* Participants must ensure that their language and actions are free of personal attacks and disparaging personal remarks.
* When interpreting the words and actions of others, participants should always assume good intentions.
* Behavior that can be reasonably considered harassment will not be tolerated.

If you don't like what other people are saying or don't like people that disagree with you, please keep your feelings to yourself.

**Reply by jose123v** [reply]
Well, it's clear that he really knows about the topic, and who is a noob with no future.

**Reply by Sophist-UK** [reply]
@jose123v 

a. Everyone starts as a noob and learns as they go. If they get things factually wrong, you can still correct them without resorting to abusive comments.

b. The whole point of being a noob is that you are likely to learn and have a better future - so your comment about their future is not only illogical, but is also against the Code of Conduct for this repo. If you don't like what other people are saying or don't like people that disagree with you, please keep your feelings to yourself.

Edit: Before calling anyone "a noob" it might actually be sensible to compare Contribution counts to make sure that it isn't actually you that is the relative noob. LoL

**Reply by jose123v** [reply]
Not even the moderators care and you are suffering so much, crystal generation 😄😄😄
If you don't like what other people are saying, please keep your feelings to yourself.

**Reply by therecluse26** [reply]
I'd personally rather laugh at these people than have them get the ban hammer. Open source has a long, storied tradition of bull-headed flame wars, it's our heritage.

**Reply by joelmellon** [reply]
> ...and who is a noob with no future.

Who's the noob? lol

**Reply by Sophist-UK** [reply]
Not @henzeb - his profile has a pretty decent contribution record.

**Comment by Sophist-UK** [comment]
@taylorotwell Can moderators please act so that this repo doesn't become an unpleasant place to contribute?

**Reply by joelmellon** [reply]
lol, I'm the reason you're not going to contribute? This thread was buried in time until a hall monitor strolled by.

**Reply by Sophist-UK** [reply]
This thread was buried in time until an off-topic email entered my email box earlier today. At which point I got annoyed by the tone of the thread and said something appropriate. I would rather not feel the need to be "a hall monitor" but when you have immature bullies throwing insults in the corridor, someone has to say something to stop the entire school becoming toxic.

**Reply by joelmellon** [reply]
Glad you recognize what you are. I like when you called for daddy's help.

**Comment by taylorotwell** [comment]
Hey all - this thread has gotten a bit out of hand. We have to keep Laravel a chill, cool place for everyone to talk and contribute. Have instituted a temporary block on a couple users to let things cool down.

**Reply by parallels999** [reply]
it would also have been great to select the correct answer to avoid confusion

**Comment by Sophist-UK** [comment]
To answer the original question, I think that the two options (kinda) already discussed are equally valid, with the final choice depending on whether you believe that this functionality would ever be included in Laravel or not:

* [Your recommended solution] Extending the class and pointing the Helper to the new version - the potential issue with this is that there are potentially hundreds of places which already use `Str::start` which now call your new version, and the stack trace if a bug in the underlying `Str` causes an exception might raise a couple of eyebrows - but so long as the calls are 100% backwards compatible I don't see any real issues with this. Your use case seems to be fully-backwards compatible to me, and the benefit of this solution is that if the Laravel version ever gets this functionality, there are no changes needed to your code to remove your own extension. The only thing you have to watch with this is the autoLoader being told to revert to the previous version on an upgrade.
* Using a macro but calling it something similar but different e.g. `Str::startMulti` for your own use - if the base version ever gets extended, then you would have to change your own code. IMO this is the best solution if you don't think the base functionality will ever be extended to include this.

**Reply by benbjurstrom** [reply]
This would be really useful for re-implementing UUID v7 for those who want it.

```php
    public function boot()
    {
        Str::macro('orderedUuid', function () {
            return Uuid::uuid7();
        });
    }
```

https://github.com/laravel/framework/pull/45546
